File: 1545-find-kth-bit-in-nth-binary-string/1545-find-kth-bit-in-nth-binary-string.go
func findKthBit(n int, k int) byte {
    s := "0"
    for i := 1; i < n; i++ {
        temp := "1" + reverse(invert(s))
        s += temp
    }

    return s[k-1]
}

// simulation, string
// time: O(2^n)
// space: O(2^n)

func reverse(s string) string {
    res := []byte(s)
    for i := 0; i < len(s)/2; i++ {
        res[i], res[len(s)-1-i] = res[len(s)-1-i], res[i]
    }
    return string(res)
}

func invert(s string) string {
    res := make([]byte, len(s))
    for i := 0; i < len(s); i++ {
        if s[i] == '1' {
            res[i] = '0'
        } else {
            res[i] = '1'
        }
    }
    return string(res)
}

File: 1545-find-kth-bit-in-nth-binary-string/1545-find-kth-bit-in-nth-binary-string.py
class Solution:
    def findKthBit(self, n: int, k: int) -> str:
        s = "0"
        for i in range(1, n):
            s += "1" + reverse(invert(s))
        return s[k - 1]

# simulation, string
# time: O(2^n)
# space: O(2^n)

def reverse(s):
    res = list(s)
    for i in range(0, len(res) // 2):
        res[i], res[len(s) - 1 - i] = res[len(s) - 1 - i], res[i]
    return "".join(res)

def invert(s):
    res = list(s)
    for i in range(0, len(res)):
        if res[i] == "1":
            res[i] = "0"
        else:
            res[i] = "1"
    return "".join(res)